src/chain-data/helpers.ts

```diff
@@ -1,23 +1,28 @@
-import produce from 'immer';
+import immerProduce, { Draft } from 'immer';
 import { ethers, providers } from 'ethers';
 import { notifications } from '../components/notifications';
 import { getDaoAddresses, getEtherscanTransactionUrl, updateNetworkName } from '../contracts';
 import { ChainData, initialChainData } from './state';
 import { convertToEnsName } from '../logic/proposals/encoding/ens-name';
 import { go } from '@api3/promise-utils';
 
-export const updateImmutably = <T>(state: T, updateCb: (immutableState: T) => void) => {
+type ProducerFn<T> = (state: Draft<T>) => void;
+
+export function produceState<T>(producerFn: ProducerFn<T>): (state: T) => T;
+export function produceState<T>(state: T, producerFn: ProducerFn<T>): T;
+export function produceState<T extends object>(stateOrProducerFn: T | ProducerFn<T>, producerFn?: ProducerFn<T>) {
```

Comment:
Btw. when I am doing the fn overloading, I just type the arguments as `any` (ususally the types just add more confusion).